# Praktikum 2
Padaa praktikum ini kita akan menggunakan library Keras untuk menggunakan JST. Keras adalah API tingkat tinggi untuk membangun JST dengan mudah, sedangkan TensorFlow adalah framework yang mendukung Keras.

Langkah:

1. Import library.

2. Load dataset.

3. Bangun model.

4. Kompilasi dan latih model.

5. Evaluasi hasil.

Klasifikasi Data Iris

In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi: {acc}")

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.2725 - loss: 1.5837   
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3113 - loss: 1.3453 
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3198 - loss: 1.1561 
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.3978 - loss: 1.1048 
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6445 - loss: 1.0531 
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6499 - loss: 1.0400 
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6256 - loss: 1.0326 
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7193 - loss: 1.0191 
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8381 - loss: 0.9908 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8622 - loss: 0.9903 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8747 - loss: 0.9756 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8714 - 

# Tugas 2

Ubah jumlah neuron hidden layer.

Bandingkan akurasi dengan konfigurasi awal.

In [ ]:
print("\n--- MODEL MODIFIKASI NEURON (20 & 16) ---")

model_neuron = tf.keras.Sequential([
    tf.keras.layers.Dense(20, activation='relu', input_shape=(4,)), # Diperbesar ke 20
    tf.keras.layers.Dense(16, activation='relu'),                   # Diperbesar ke 16
    tf.keras.layers.Dense(3, activation='softmax')
])

model_neuron.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history_neuron = model_neuron.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

loss_neuron, acc_neuron = model_neuron.evaluate(X_test, y_test, verbose=0)
print(f"Modifikasi Neuron - Loss: {loss_neuron:.4f}, Akurasi: {acc_neuron:.4f}")

# Bandingkan
if acc_neuron > acc:
    print("Result: Menambah neuron meningkatkan akurasi.")
else:
    print("Result: Menambah neuron tidak berpengaruh signifikan/menurunkan performa (overfitting).")


--- MODEL MODIFIKASI NEURON (20 & 16) ---


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Modifikasi Neuron - Loss: 0.1337, Akurasi: 1.0000
Result: Menambah neuron tidak berpengaruh signifikan/menurunkan performa (overfitting).


mengubah jumlah neuron pada hidden layer menjadi lebih besar (misalnya 2x lipat) untuk melihat apakah menambah "kapasitas otak" model akan meningkatkan akurasi atau justru tidak berpengaruh.

Layer 1: Ubah 10 menjadi 20

Layer 2: Ubah 8 menjadi 16

# Tugas 3

Bandingkan Sigmoid vs ReLU pada dataset Iris.

Catat perbedaan loss dan akurasi.

In [ ]:
# --- TUGAS 3: GANTI RELU DENGAN SIGMOID ---
print("\n--- MODEL AKTIVASI SIGMOID ---")

model_sigmoid = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='sigmoid', input_shape=(4,)), # Ganti ke Sigmoid
    tf.keras.layers.Dense(8, activation='sigmoid'),                    # Ganti ke Sigmoid
    tf.keras.layers.Dense(3, activation='softmax')
])

model_sigmoid.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history_sigmoid = model_sigmoid.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)

loss_sigmoid, acc_sigmoid = model_sigmoid.evaluate(X_test, y_test, verbose=0)
print(f"Sigmoid - Loss: {loss_sigmoid:.4f}, Akurasi: {acc_sigmoid:.4f}")
print(f"Baseline (ReLU) - Loss: {loss:.4f}, Akurasi: {acc:.4f}")


--- MODEL AKTIVASI SIGMOID ---
Sigmoid - Loss: 0.7807, Akurasi: 0.6000
Baseline (ReLU) - Loss: 0.2757, Akurasi: 1.0000


Mengganti fungsi aktivasinya dari ReLU menjadi Sigmoid pada hidden layer.

### 1. Analisis Baseline (Model Awal)
* **Hasil:** Akurasi 1.0 (100%) | Loss: 0.2757
* **Penjelasan:**
    * Dataset Iris tergolong dataset sederhana. Dengan konfigurasi awal (10 neuron layer 1, 8 neuron layer 2), model sudah memiliki kapasitas yang cukup untuk memisahkan ke-3 jenis bunga dengan sempurna.
    * **Pesan Warning:** `UserWarning: Do not pass an input_shape...` adalah peringatan bahwa sintaks Keras yang baru lebih menyukai penggunaan `Input(shape)` terpisah. Ini **tidak mempengaruhi hasil** (hanya masalah gaya penulisan kode).

---

### 2. Analisis Tugas 2 (Modifikasi Neuron: 20 & 16)
* **Hasil:** Akurasi 1.0 (100%) | Loss: **0.1337** (Jauh lebih rendah dari Baseline 0.2757)
* **Koreksi Kesimpulan Code:**
    * Meskipun kode mencetak *"Menambah neuron tidak berpengaruh signifikan"*, itu hanya karena akurasi mentok di 100% (tidak bisa lebih tinggi lagi).
    * **Fakta Sebenarnya:** Performa meningkat! Lihat nilai **Loss**-nya.
    * **Loss 0.27 vs 0.13:**
        * Loss **0.27** (Baseline): Model menebak benar, tapi mungkin tingkat keyakinannya sekitar 70-80%.
        * Loss **0.13** (Modifikasi): Model menebak benar dengan tingkat keyakinan mendekati 90-95%.
    * **Kesimpulan:** Menambah jumlah neuron membuat model menjadi lebih "pintar" dan **lebih percaya diri** (confident) dengan tebakannya, meskipun jumlah tebakan benarnya sama-sama 100%.

---

### 3. Analisis Tugas 3 (Sigmoid vs ReLU)
Ini adalah bagian paling penting.
* **Hasil Sigmoid:** Akurasi 0.60 (60%) | Loss: 0.7807
* **Hasil ReLU:** Akurasi 1.0 (100%) | Loss: 0.2757
* **Mengapa Sigmoid Hancur (Drop 40%)?**
    1.  **Vanishing Gradient Problem:** Pada fungsi Sigmoid, turunan (gradient) di ujung-ujung kurva nilainya sangat kecil (mendekati 0). Ketika melakukan *backpropagation* (perambatan mundur), nilai update bobot menjadi sangat kecil. Akibatnya, model belajar **sangat lambat**.
    2.  **Epoch Terbatas:** hanya melatih selama 50 epoch. Untuk ReLU, 50 epoch sudah lebih dari cukup. Untuk Sigmoid, 50 epoch belum cukup baginya untuk menemukan pola yang optimal karena lambatnya proses belajar tadi.
    3.  **Saturasi:** Output Sigmoid terjepit antara 0 dan 1. Jika inisialisasi bobot tidak hati-hati, neuron bisa cepat "jenuh" (saturated) dan berhenti belajar.

### Kesimpulan Akhir
Dari eksperimen ini, kita bisa menyimpulkan:

1.  **ReLU Superior:** Untuk *hidden layer*, ReLU hampir selalu lebih baik daripada Sigmoid karena proses belajarnya lebih cepat dan terhindar dari masalah gradien yang hilang (*vanishing gradient*).
2.  **Kapasitas Model:** Menambah neuron (Task 2) menurunkan *loss* (error), yang berarti model semakin presisi memisahkan data, meskipun akurasinya sudah maksimal.
3.  **Dataset Iris Mudah:** Karena dataset ini sederhana, model kecil (Baseline) pun sudah bisa mencapai akurasi 100%.

**Saran:** Gunakan selalu **ReLU** untuk hidden layer, dan gunakan **Sigmoid/Softmax** hanya di output layer (tergantung binary atau multi-class).